File: Algorithms/0238_Product_of_Array_Except_Self/Python/Product_of_Array_Except_Self_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def productExceptSelf(self, nums):
        list1 = [1 for _ in range(len(nums))]
        list2 = [1 for _ in range(len(nums))]

        for i in range(1, len(nums)):
            list1[i] = list1[i - 1] * nums[i - 1]
        for i in range(len(nums) - 2, -1, -1):
            list2[i] = list2[i + 1] * nums[i + 1]

        return [i * j for i, j in zip(list1, list2)]

File: Algorithms/0238_Product_of_Array_Except_Self/Java/ProductofArrayExceptSelfSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int[] productExceptSelf(int[] nums) {
        int[] temp1 = new int[nums.length];
        int[] temp2 = new int[nums.length];
        int[] res = new int[nums.length];
        Arrays.fill(temp1, 1);
        Arrays.fill(temp2, 1);

        for (int i = 1; i < nums.length; i++) {
            temp1[i] = temp1[i - 1] * nums[i - 1];
        }
        for (int i = nums.length - 2; i >= 0; i--) {
            temp2[i] = temp2[i + 1] * nums[i + 1];
        }
        for (int i = 0; i < res.length; i++) {
            res[i] = temp1[i] * temp2[i];
        }
        return res;
    }
}




